BARK-311: Least-privilege collection documentation for AzureHound (#245)
SpecterOps/bloodhound-docs: docs/install-data-collector/install-azurehound/azure-configuration.mdx

Reviewer: ```suggestion
<Note>All steps in this section are optional and do not affect the functionality of the collector.</Note>
```

description: "This section details creating and configuring an Enterprise Application for AzureHound within Microsoft Entra ID, including API permissions, roles, and authentication certificate."
---

import EntraIdPermissions from '/snippets/hounds/azurehound-entra-id-permissions.mdx';

<img noZoom src="/assets/enterprise-edition-pill-tag.svg" alt="Applies to BloodHound Enterprise only"/>

> <Info>You can complete the steps on this page manually following the instructions below, or use the PowerShell script at the [Scripted Configuration](#scripted-configuration) section to automate the entire process (except certificate upload).</Info>

Reviewer: ```suggestion
<Info>You can complete the steps on this page manually following the instructions below, or use the PowerShell script in the [Scripted Configuration](#scripted-configuration) section to automate the entire process (except certificate upload).</Info>
```

<img noZoom src="/assets/enterprise-edition-pill-tag.svg" alt="Applies to BloodHound Enterprise only"/>

> <Info>You can complete the steps on this page manually using the sections below, or use the PowerShell script at the [Scripted Configuration](#scripted-configuration) section to automate the entire process (except certificate upload).</Info>

Reviewer 2: Instead of "using the sections below", I would prefer "following the instructions below" or something in taht sense.

      <img src="/assets/image-109.png" alt=""/>
    </Frame>

> 6.  Search for all MS Graph permissions described in [AzureHound Data and Permissions: Entra ID](/collect-data/azurehound-data-permissions#entra-id) and check the box next to them.

Reviewer 2: I would probably still list the permissions here as well, with a hyperlink to the details, so that the user does not need to navigate back and forth when following the guide.

Author: The current way avoids maintaining two articles. I will make a snippet instead. Thanks!

      <img src="/assets/image-109.png" alt=""/>
    </Frame>

6.  Search for and check the box next to each of the following Microsoft Graph **application** permissions. See [AzureHound Data and Permissions: Entra ID](/collect-data/azurehound-data-permissions#entra-id) for details on the least-privilege approach:

> <EntraIdPermissions />

Reviewer: ```suggestion
    <EntraIdPermissions />
```

Reviewer: Increasing the indentation will align the table with the text of the ordered list. Otherwise, it looks like this:

<img width="984" height="503" alt="Image" src="https://github.com/user-attachments/assets/e6e07f96-efc7-4028-941d-9a5ae8d811fe" />

7.  In the bottom of the window, select **Add permissions**.
    <Frame>
      <img src="/assets/image-111.png" alt=""/>
    </Frame>

8.  Click on **Grant admin consent for &lt;your\_tenant\_name&gt;**.
    <Frame>
      <img src="/images/data_collectors/grant_admin_consent.png"/>
    </Frame>

9.  Click **Yes** on the confirmation dialog.
    <Frame>
      <img src="/assets/image-113.png" alt=""/>
    </Frame>

10.  After being redirected to API Permissions again, you should see the **Status** column listing all permissions as **Granted**.

>     <Frame>

Reviewer: ```suggestion
     <Frame>
```

>       <img src="/assets/azurehound-install-consent-granted.png" alt=""/>

Reviewer: ```suggestion
       <img src="/assets/azurehound-install-consent-granted.png" alt=""/>
```

>     </Frame>

Reviewer: ```suggestion
     </Frame>
```

      <img src="/assets/azurehound-install-branding-result.png" alt=""/>
    </Frame>

7. Continue to the next section to create and assign a custom role in Azure Resource Manager

## Create and assign custom AzureHound Reader role in Azure Resource Manager

> _Note: If you do not have any management groups, you may either create your Tenant Root Group following the prompts in the middle of the screen to ensure future visibility if another administrator begins use of subscriptions, or you may skip this section altogether. If you skip this section, you will see a warning in the logs for each collection indicating the lack of ability to collect this data accordingly._

Reviewer: ```suggestion
<Note>
  If you don't have any management groups, you can skip this section. However, AzureHound will log a warning during each collection indicating it cannot collect management group data.

  Alternatively, you can create your Tenant Root Group by following the prompts in the Azure portal. This ensures visibility if another administrator begins using subscriptions in the future.
</Note>
```

2.  Search for and select the **Management groups** item in the top search bar.
    <Frame>
      <img src="/assets/arm-search-management-groups.png" alt=""/>
    </Frame>

3.  Select **Tenant Root Group**.
    <Frame>
      <img src="/assets/arm-select-tenant-root-group.png" alt=""/>
    </Frame>

4.  Select **Access control (IAM)**.
    <Frame>
      <img src="/assets/arm-select-access-control-iam.png" alt=""/>
    </Frame>

5.  Click **Add**, then **Add custom role**.
    <Frame>
      <img src="/assets/arm-add-custom-role.png" alt=""/>
    </Frame>

6.  Download  <a href="/assets/azurehound-reader-role.json" download>azurehound-reader-role.json</a>, open it in a text editor, and replace `<tenantRootGroupId>` with your [Tenant Root Management Group ID](https://learn.microsoft.com/en-us/azure/governance/management-groups/overview#root-management-group-for-each-directory) (this is your Entra ID tenant ID).
    <Frame>
      <img src="/assets/arm-json-tenantid.png" alt=""/>
    </Frame>

> 

Reviewer: The screenshot is totally fine, but if you wanted to use a code block instead, you could do something like this:

```
    ```json azurehound-reader-role.json lines highlight={6}
    {
        "properties": {
            "roleName": "AzureHound Reader",
            "description": "Least-privilege read-only role for AzureHound data collection from Azure Resource Manager.",
            "assignableScopes": [
                "/providers/Microsoft.Management/managementGroups/<tenantRootGroupId>"
            ],
    ```
```

Or:

```
    ```json azurehound-reader-role.json lines focus={6}
    {
        "properties": {
            "roleName": "AzureHound Reader",
            "description": "Least-privilege read-only role for AzureHound data collection from Azure Resource Manager.",
            "assignableScopes": [
                "/providers/Microsoft.Management/managementGroups/<tenantRootGroupId>"
            ],
    ```
```

7. In **Basics** > **File**, upload the edited file and click **Review + create**.
    <Frame>
      <img src="/assets/arm-upload-azurehound-reader.png" alt=""/>
    </Frame>

<Information>See [AzureHound Data Collection and Permissions: Azure Resource Manager](/collect-data/azurehound-data-permissions#azure-resource-manager) for details on each least-privilege permission in the AzureHound Reader role.</Information>

8. Review the role and click **Create** at the bottom of the page.
    <Frame>
      <img src="/assets/arm-create.png" alt=""/>
    </Frame>

9. Back in **Access control (IAM)**, click **Add**, then **Add role assignment**.
    <Frame>
      <img src="/assets/arm-add-role-assignment.png" alt=""/>
    </Frame>

10.  Search for the **AzureHound Reader** role and select it.

>     <Frame>

Reviewer: ```suggestion
     <Frame>
```

>       <img src="/assets/arm-select-azurehound-reader-role.png" alt=""/>

Reviewer: ```suggestion
       <img src="/assets/arm-select-azurehound-reader-role.png" alt=""/>
```

>     </Frame>

Reviewer: ```suggestion
     </Frame>
```

11.  Click **Members**.

>     <Frame>

Reviewer: ```suggestion
     <Frame>
```

>       <img src="/assets/image-130.png" alt=""/>

Reviewer: ```suggestion
       <img src="/assets/image-130.png" alt=""/>
```

>     </Frame>

Reviewer: ```suggestion
     </Frame>
```

12.  Click **Select members**.

>     <Frame>

Reviewer: ```suggestion
     <Frame>
```

>       <img src="/assets/image-131.png" alt=""/>

Reviewer: ```suggestion
       <img src="/assets/image-131.png" alt=""/>
```

>     </Frame>

Reviewer: ```suggestion
     </Frame>
```

13.  Search for and click on your previously created service principal.

>     <Frame>

Reviewer: ```suggestion
     <Frame>
```

>       <img src="/assets/image-132.png" alt=""/>

Reviewer: ```suggestion
       <img src="/assets/image-132.png" alt=""/>
```

>     </Frame>

Reviewer: ```suggestion
     </Frame>
```

14. Validate the principal selected, then click **Select**.

>     <Frame>

Reviewer: ```suggestion
     <Frame>
```

>       <img src="/assets/image-133.png" alt=""/>

Reviewer: ```suggestion
       <img src="/assets/image-133.png" alt=""/>
```

>     </Frame>

Reviewer: ```suggestion
     </Frame>
```

15. Click the tab **Review + Assign**.

>     <Frame>

Reviewer: ```suggestion
     <Frame>
```

>       <img src="/assets/image-134.png" alt=""/>

Reviewer: ```suggestion
       <img src="/assets/image-134.png" alt=""/>
```

>     </Frame>

Reviewer: ```suggestion
     </Frame>
```

16. Click **Review + Assign** at the bottom of the page.

>     <Frame>

Reviewer: ```suggestion
     <Frame>
```

>       <img src="/assets/image-135.png" alt=""/>

Reviewer: ```suggestion
       <img src="/assets/image-135.png" alt=""/>
```

>     </Frame>

Reviewer: ```suggestion
     </Frame>
```

17. Confirm the role is present by refreshing this view. You may need to alter the filter to see this role.

18. Continue to [Run and Upgrade AzureHound (Windows, Docker, or Kubernetes)](/install-data-collector/install-azurehound/installation-options)

## Scripted configuration

As an alternative to the manual steps described in this document, you can use **PowerShell** to automate the entire configuration process.
The script below registers AzureHound in Entra ID, creates the least-privilege Azure Reader role, and assigns it to the service principal.
The script is **idempotent**—it safely handles re-runs by detecting existing resources and updating them as needed, preventing duplicate role definitions or assignments.

> After running this script, follow the [Add application authentication certificate](#add-application-authentication-certificate) section to complete the setup.

Reviewer: ```suggestion
As an alternative to the manual steps described in this document, you can use **PowerShell** to automate the entire configuration process.

The script below registers AzureHound in Entra ID, creates the least-privilege Azure Reader role, and assigns it to the service principal.
The script is **idempotent**—it safely handles re-runs by detecting existing resources and updating them as needed, preventing duplicate role definitions or assignments.

<Note>
  After running this script, follow the [Add application authentication certificate](#add-application-authentication-certificate) section to complete the setup.
</Note>
```

> ```powershell

Reviewer: ```suggestion
```powershell ConfigureAzureHound.ps1
```

2.  Search for and select the **Management groups** item in the top search bar.
    <Frame>
      <img src="/assets/arm-search-management-groups.png" alt=""/>
    </Frame>

3.  Select **Tenant Root Group**.
    <Frame>
      <img src="/assets/arm-select-tenant-root-group.png" alt=""/>
    </Frame>

4.  Select **Access control (IAM)**.
    <Frame>
      <img src="/assets/arm-select-access-control-iam.png" alt=""/>
    </Frame>

5.  Click **Add**, then **Add custom role**.
    <Frame>
      <img src="/assets/arm-add-custom-role.png" alt=""/>
    </Frame>

6.  Download  <a href="/assets/azurehound-reader-role.json" download>azurehound-reader-role.json</a>, open it in a text editor, and replace `<tenantRootGroupId>` with your [Tenant Root Management Group ID](https://learn.microsoft.com/en-us/azure/governance/management-groups/overview#root-management-group-for-each-directory) (this is your Entra ID tenant ID).
    <Frame>
      <img src="/assets/arm-json-tenantid.png" alt=""/>
    </Frame>

7. In **Basics** > **File**, upload the edited file and click **Review + create**.
    <Frame>
      <img src="/assets/arm-upload-azurehound-reader.png" alt=""/>
    </Frame>

> <Information>See [AzureHound Data Collection and Permissions: Azure Resource Manager](/collect-data/azurehound-data-permissions#azure-resource-manager) for details on each least-privilege permission in the AzureHound Reader role.</Information>

Reviewer: ```suggestion
    <Info>See [AzureHound Data Collection and Permissions: Azure Resource Manager](/collect-data/azurehound-data-permissions#azure-resource-manager) for details on each least-privilege permission in the AzureHound Reader role.</Info>
```

Reviewer: This isn't rendering because `Information` isn't a recognized component. It also needs to be indented.

## Scripted configuration

15. Continue to [Run and Upgrade AzureHound (Windows, Docker, or Kubernetes)](/install-data-collector/install-azurehound/installation-options)
As an alternative to the manual steps described in this document, you can use **PowerShell** to automate the entire configuration process.

## Scripted app registration
The script below registers AzureHound in Entra ID, creates the least-privilege Azure Reader role, and assigns it to the service principal.
The script is **idempotent**—it safely handles re-runs by detecting existing resources and updating them as needed, preventing duplicate role definitions or assignments.

As an alternative to the manual app registration process described in this document, you can use **PowerShell** to achieve the same.
Below is a sample PowerShell script that performs all the required steps,
with the exception of [configuring the authentication certificate](#add-application-authentication-certificate).
<Note>
  After running this script, follow the [Add application authentication certificate](#add-application-authentication-certificate) section to complete the setup.
</Note>

> ```powershell

Reviewer: I think this was an inadvertent deletion. It needs to be re-applied.

I don't know why the Mintlify PR check isn't flagging it, but if you run mint validate locally you'll see the error:

```
mint validate

parsing error ./install-data-collector/install-azurehound/azure-configuration.mdx:259:2 - Unexpected character `#` (U+0023) before name, expected a character that can start a name, such as a letter, `$`, or `_`
warning - "install-data-collector/install-azurehound/azure-configuration" is referenced in the docs.json navigation but the file does not exist.
error Build validation failed with 1 warning(s). See above for details.
```

.DESCRIPTION

> This script registers the Azure Data Exporter for BloodHound Enterprise (AzureHound)

Reviewer: Should the script include a SpecterOps copyright and/or licensing header?

Author: I am unsure. I did not create the script, @MichaelGrafnetter did.

Reviewer 2: As the original author of the script, I am not sure either. Perhaps we could at least add `Authors:` to the `.NOTES` section and list our names and company. And `License: Apache License, Version 2.0`? `Version: 1.1` could probably be increased as well. But I would prefer not to clutter the script with a full licensing header.

[Microsoft.Graph.PowerShell.Models.IMicrosoftGraphServicePrincipal] $microsoftGraph =
    Get-MgServicePrincipal -Filter "DisplayName eq 'Microsoft Graph'"

# Define the least-privilege Microsoft Graph application permissions
# See: /collect-data/azurehound-data-permissions#entra-id
[string[]] $requiredPermissions = @(
    'User.Read.All'
    'GroupMember.Read.All'
    'Application.Read.All'
    'Device.Read.All'
    'Organization.Read.All'
    'RoleManagement.Read.Directory'
    'AdministrativeUnit.Read.All'
    'AuditLog.Read.All'  # Optional: for signInActivity on user objects
)

# Resolve each permission name to its AppRole definition

> [hashtable[]] $resourceAccess = $requiredPermissions | ForEach-Object {

Reviewer 2: I prefer `$PSItem` to `$_` 😉
Could the variable type be `[Microsoft.Graph.PowerShell.Models.IMicrosoftGraphAppRole[]]` instead of `[hashtable[]]`?

Author: `$PSItem` 👍 

Regarding the second, you would know better than me. Claude's opinion makes sense to me though:

> [hashtable[]] is actually correct here. The code outputs hashtables with { id = $role.Id; type = 'Role' }, not IMicrosoftGraphAppRole objects. The type [Microsoft.Graph.PowerShell.Models.IMicrosoftGraphAppRole[]] would only work if you were returning $role directly. That said, the type annotation itself is technically optional—PowerShell will infer the array type from the output. If you wanted to keep it clean and less prescriptive, you could remove it entirely and let the array type be inferred.

Reviewer 2: Perhaps splitting it first to a `[Microsoft.Graph.PowerShell.Models.IMicrosoftGraphAppRole[]]` variable and then to `[hashtable[]]` variable, to make it more readable.

#endregion Entra ID

#region Azure

# Optionally enable browser-based login on Windows 10 and later
Update-AzConfig -EnableLoginByWam $false

# Authenticate against Azure Resource Manager
Connect-AzAccount -Environment AzureCloud -Scope Process

# Download the AzureHound Reader custom role definition
[string] $roleDefinitionUrl =

>   'https://bloodhound.specterops.io/assets/azurehound-reader-role.json'

Reviewer 2: I would also consider embedding the role definition directly in the script using here string syntax. Just an idea.

Author: The consequence is double maintenance and risk of drift. @jeff-matthews, you have an opinion?

Reviewer 2: I agree with your point of view. It's just that the JSON file contains security-sensitive config.

Author: Gotcha. I also see that MS Graph permissions are hardcoded anyway too in `$requiredPermissions`. Perhaps we should stick to that

Reviewer: @martinsohn, it's good instinct ordinarily, but in this case I think you're right to replace the file reference as suggested. There's probably a way to semi-automate as a pre-processing step with a `just` recipe, but I don't think we need to worry about it right now.

[Microsoft.Azure.Commands.Resources.Models.ManagementGroups.PSManagementGroup] $rootManagementGroup =
    Get-AzManagementGroup -GroupName $currentTenantId

# New-AzRoleDefinition expects a flat structure, not the nested `properties` wrapper
# used by the ARM portal upload format, so we must flatten it before passing it in.
$roleDefinition = Get-Content -Path $tempRolePath -Raw | ConvertFrom-Json
Remove-Item -Path $tempRolePath

> 

Reviewer 2: $roleDefinition, $flatRole etc. do not have types, e.g., [psobject] or [PSCustomObject].

    
    if ($existingRole) {
        $existingRole.Actions = $flatRole.Actions
        $existingRole.NotActions = $flatRole.NotActions
        $existingRole.DataActions = $flatRole.DataActions
        $existingRole.NotDataActions = $flatRole.NotDataActions
        $existingRole.Description = $flatRole.Description
        $existingRole.AssignableScopes = $flatRole.AssignableScopes
        [Microsoft.Azure.Commands.Resources.Models.Authorization.PSRoleDefinition] $azureHoundReaderRole =
            Set-AzRoleDefinition -Role $existingRole
    }
    else {
        [Microsoft.Azure.Commands.Resources.Models.Authorization.PSRoleDefinition] $azureHoundReaderRole =
            New-AzRoleDefinition -InputFile $flatRolePath
    }
}
finally {
    Remove-Item -Path $flatRolePath
}

# Assign the AzureHound Reader role at the Tenant Root Group
$existingAssignment = Get-AzRoleAssignment -ObjectId $servicePrincipal.Id `
                                           -Scope $rootManagementGroup.Id `
                                           -RoleDefinitionId $azureHoundReaderRole.Id `
                                           -ErrorAction SilentlyContinue

> if ($existingAssignment) {

Reviewer 2: I'd prefer `$null -ne $existingRoleAssignment`
